File: ros2_ws/src/lerobot_controller/scripts/joint_state_to_trajectory.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import JointState
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint
from builtin_interfaces.msg import Duration


class JointStateToTrajectory(Node):
    def __init__(self):
        super().__init__('joint_state_to_trajectory')

        # Parameters
        self.declare_parameter('input_topic', '/leader/joint_states')
        self.declare_parameter('arm_output_topic', '/follower/arm_controller/joint_trajectory')
        self.declare_parameter('gripper_output_topic', '/follower/gripper_controller/joint_trajectory')
        self.declare_parameter('arm_joints', ['1', '2', '3', '4', '5'])
        self.declare_parameter('gripper_joints', ['6'])
        self.declare_parameter('time_from_start_sec', 0.1)

        input_topic = self.get_parameter('input_topic').value
        arm_output_topic = self.get_parameter('arm_output_topic').value
        gripper_output_topic = self.get_parameter('gripper_output_topic').value
        self.arm_joints = self.get_parameter('arm_joints').value
        self.gripper_joints = self.get_parameter('gripper_joints').value
        time_from_start = self.get_parameter('time_from_start_sec').value

        self.time_from_start = Duration(sec=int(time_from_start), nanosec=int((time_from_start % 1) * 1e9))

        # Subscriber
        self.subscription = self.create_subscription(
            JointState,
            input_topic,
            self.joint_state_callback,
            10
        )

        # Publishers
        self.arm_pub = self.create_publisher(
            JointTrajectory,
            arm_output_topic,
            10
        )
        self.gripper_pub = self.create_publisher(
            JointTrajectory,
            gripper_output_topic,
            10
        )

        self.get_logger().info(f'Joint state to trajectory converter started')
        self.get_logger().info(f'  Input: {input_topic}')
        self.get_logger().info(f'  Arm output: {arm_output_topic}')
        self.get_logger().info(f'  Gripper output: {gripper_output_topic}')
        self.get_logger().info(f'  Arm joints: {self.arm_joints}')
        self.get_logger().info(f'  Gripper joints: {self.gripper_joints}')

    def joint_state_callback(self, msg):
        # Create a mapping from joint name to position
        joint_dict = {}
        for i, name in enumerate(msg.name):
            if i < len(msg.position):
                joint_dict[name] = msg.position[i]

        # Publish arm trajectory
        arm_positions = []
        for joint in self.arm_joints:
            if joint in joint_dict:
                arm_positions.append(joint_dict[joint])
            else:
                self.get_logger().warn(f'Joint {joint} not found in joint_states', throttle_duration_sec=5.0)
                return

        if len(arm_positions) == len(self.arm_joints):
            arm_traj = JointTrajectory()
            arm_traj.header.stamp = self.get_clock().now().to_msg()
            arm_traj.joint_names = self.arm_joints

            point = JointTrajectoryPoint()
            point.positions = arm_positions
            point.time_from_start = self.time_from_start
            arm_traj.points = [point]

            self.arm_pub.publish(arm_traj)

        # Publish gripper trajectory
        gripper_positions = []
        for joint in self.gripper_joints:
            if joint in joint_dict:
                gripper_positions.append(joint_dict[joint])
            else:
                self.get_logger().warn(f'Joint {joint} not found in joint_states', throttle_duration_sec=5.0)
                return

        if len(gripper_positions) == len(self.gripper_joints):
            gripper_traj = JointTrajectory()
            gripper_traj.header.stamp = self.get_clock().now().to_msg()
            gripper_traj.joint_names = self.gripper_joints

            point = JointTrajectoryPoint()
            point.positions = gripper_positions
            point.time_from_start = self.time_from_start
            gripper_traj.points = [point]

            self.gripper_pub.publish(gripper_traj)


def main(args=None):
    rclpy.init(args=args)
    node = JointStateToTrajectory()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2_ws/src/lerobot_controller/src/joint_state_to_trajectory.cpp
#include "lerobot_controller/joint_state_to_trajectory.hpp"
#include <chrono>

using namespace std::chrono_literals;

namespace lerobot_controller {

JointStateToTrajectory::JointStateToTrajectory()
    : Node("joint_state_to_trajectory") {

  // Declare parameters
  this->declare_parameter<std::string>("input_topic", "/leader/joint_states");
  this->declare_parameter<std::string>("arm_output_topic", "/follower/arm_controller/joint_trajectory");
  this->declare_parameter<std::string>("gripper_output_topic", "/follower/gripper_controller/joint_trajectory");
  this->declare_parameter<std::vector<std::string>>("arm_joints", {"1", "2", "3", "4", "5"});
  this->declare_parameter<std::vector<std::string>>("gripper_joints", {"6"});
  this->declare_parameter<double>("time_from_start_sec", 0.1);

  // Get parameters
  auto input_topic = this->get_parameter("input_topic").as_string();
  auto arm_output_topic = this->get_parameter("arm_output_topic").as_string();
  auto gripper_output_topic = this->get_parameter("gripper_output_topic").as_string();
  arm_joints_ = this->get_parameter("arm_joints").as_string_array();
  gripper_joints_ = this->get_parameter("gripper_joints").as_string_array();
  time_from_start_sec_ = this->get_parameter("time_from_start_sec").as_double();

  // Create subscriber
  joint_state_sub_ = this->create_subscription<sensor_msgs::msg::JointState>(
      input_topic,
      10,
      std::bind(&JointStateToTrajectory::joint_state_callback, this, std::placeholders::_1));

  // Create publishers
  arm_pub_ = this->create_publisher<trajectory_msgs::msg::JointTrajectory>(arm_output_topic, 10);
  gripper_pub_ = this->create_publisher<trajectory_msgs::msg::JointTrajectory>(gripper_output_topic, 10);

  RCLCPP_INFO(this->get_logger(), "Joint state to trajectory converter started");
  RCLCPP_INFO(this->get_logger(), "  Input: %s", input_topic.c_str());
  RCLCPP_INFO(this->get_logger(), "  Arm output: %s", arm_output_topic.c_str());
  RCLCPP_INFO(this->get_logger(), "  Gripper output: %s", gripper_output_topic.c_str());
}

void JointStateToTrajectory::joint_state_callback(const sensor_msgs::msg::JointState::SharedPtr msg) {
  // Create a mapping from joint name to position
  std::unordered_map<std::string, double> joint_dict;
  for (size_t i = 0; i < msg->name.size() && i < msg->position.size(); ++i) {
    joint_dict[msg->name[i]] = msg->position[i];
  }

  // Publish arm trajectory
  std::vector<double> arm_positions;
  arm_positions.reserve(arm_joints_.size());
  bool arm_complete = true;
  for (const auto& joint : arm_joints_) {
    auto it = joint_dict.find(joint);
    if (it != joint_dict.end()) {
      arm_positions.push_back(it->second);
    } else {
      RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 5000,
                           "Joint %s not found in joint_states", joint.c_str());
      arm_complete = false;
      break;
    }
  }

  if (arm_complete && arm_positions.size() == arm_joints_.size()) {
    auto arm_traj = trajectory_msgs::msg::JointTrajectory();
    arm_traj.header.stamp = this->now();
    arm_traj.joint_names = arm_joints_;

    auto point = trajectory_msgs::msg::JointTrajectoryPoint();
    point.positions = arm_positions;
    point.time_from_start.sec = static_cast<int32_t>(time_from_start_sec_);
    point.time_from_start.nanosec = static_cast<uint32_t>((time_from_start_sec_ - static_cast<int32_t>(time_from_start_sec_)) * 1e9);
    arm_traj.points.push_back(point);

    arm_pub_->publish(arm_traj);
  }

  // Publish gripper trajectory
  std::vector<double> gripper_positions;
  gripper_positions.reserve(gripper_joints_.size());
  bool gripper_complete = true;
  for (const auto& joint : gripper_joints_) {
    auto it = joint_dict.find(joint);
    if (it != joint_dict.end()) {
      gripper_positions.push_back(it->second);
    } else {
      RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 5000,
                           "Joint %s not found in joint_states", joint.c_str());
      gripper_complete = false;
      break;
    }
  }

  if (gripper_complete && gripper_positions.size() == gripper_joints_.size()) {
    auto gripper_traj = trajectory_msgs::msg::JointTrajectory();
    gripper_traj.header.stamp = this->now();
    gripper_traj.joint_names = gripper_joints_;

    auto point = trajectory_msgs::msg::JointTrajectoryPoint();
    point.positions = gripper_positions;
    point.time_from_start.sec = static_cast<int32_t>(time_from_start_sec_);
    point.time_from_start.nanosec = static_cast<uint32_t>((time_from_start_sec_ - static_cast<int32_t>(time_from_start_sec_)) * 1e9);
    gripper_traj.points.push_back(point);

    gripper_pub_->publish(gripper_traj);
  }
}

}  // namespace lerobot_controller

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<lerobot_controller::JointStateToTrajectory>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
